### DISCUSSION TITLE
How to intercept events for any control and restore the event propagation within that control

### DISCUSSION THREAD

**Discussion by wobushiafa** [AvaloniaUI/Avalonia#14045] [General]
I have customized a control that listens to mouse scrolling events and pinch gestures. After placing it inside a ScrollViewer, I want to add a flag to prevent the ScrollViewer from scrolling and handling gestures. Instead, I aim to redirect the events back to my custom control. However, I am currently stuck on how to pass the scroll and pinch events to my custom control

**Comment by thevortexcloud** [comment]
Have you looked at the event routing docs?

https://docs.avaloniaui.net/docs/concepts/input/routed-events

Typically you can do this with a Tunneling event and just not set the event as handled. In which case it will keep propagating.

**Comment by wobushiafa** [comment]
> 您是否查看过事件路由文档？
> 
> https://docs.avaloniaui.net/docs/concepts/input/routed-events
> 
> 通常，您可以使用 Tunneling 事件执行此操作，但不要将事件设置为已处理。在这种情况下，它将继续传播。

By default, the ScrollViewer in scrolls its content when the mouse wheel is scrolled.   Additionally, the inner controls also propagate the scroll events.   Now, I want to introduce a flag.   When the flag is set to true, mouse scrolling should only affect the outer ScrollViewer, causing a change in the content viewport while preventing the inner controls from triggering scroll events.   When the flag is set to false, mouse scrolling should not affect the outer ScrollViewer's viewport, but the inner controls should still trigger scroll events.   I attempted to listen to the mouse scroll event on the inner controls and successfully consumed the bubbling event when the flag was set to false.   It worked as expected.   However, when the flag is set to true, scrolling the mouse causes the ScrollViewer's content viewport to scroll, and at the same time, the inner controls also respond to the scroll events.

**Reply by timunie** [reply]
You may need to tunnel the event and mark it as handled in event args

**Reply by wobushiafa** [reply]
tunnel ScrollViewer or other control?  I feel it may not be feasible.  What I need to achieve is that in any situation, either the inner controls respond to the event, or the outer ScrollViewer responds to the event.  Both of them cannot respond to the event simultaneously

**Reply by timunie** [reply]
Check ScrollViewer code. It also only scrolls the inner sv first and then the outer. But your inner control needs to mark it as handled. 

numericUpDown also may be a good candidate for checking code

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/Avalonia/blob/master/src%2FAvalonia.Controls%2FButtonSpinner.cs#L193-L207

I guess that's what you need

**Reply by wobushiafa** [reply]
First and foremost, thank you very much for your assistance,The task I am currently working on involves encapsulating a ZoomContentControl that responds to PointerWheel events, PointerMoved events, and pinch gestures to implement content zooming and panning.   The current business requirement is to wrap it with an additional ScrollViewer externally.   After reviewing the information you provided, I'm not sure if I truly understand the implications.   Within the ZoomContentControl, I have defined a StyledProperty, and when it is set to true, the PointerWheel event it listens to will set e.Handle=true.   I have successfully achieved the requirement of preventing scrolling of the viewport externally by using this property.   However, when setting this StyledProperty to false, I'm still unsure how to implement scrolling only for the ScrollViewer.   Do I need to make changes to the ZoomContentControl?   Placing ZoomContentControl inside the ScrollViewer was a reluctant necessity, and I hope to make ZoomContentControl more versatile.   Additionally, ScrollViewer supports ScrollGestureRecognizer, which conflicts with ZoomContentControl's PinchGestureRecognizer.   I'm unsure how to disable one of them.

**Reply by timunie** [reply]
There is a PanAndZoom control available. Are you aware of it? If not search it on GitHub

**Reply by wobushiafa** [reply]
No, his functionality is not aligned with my goals. The current requirement is that the content size is already determined to be larger than the screen size. Therefore, it needs to be nested within a ScrollViewer to allow scrolling and viewing all content. Subsequently, it should be possible to zoom and pan based on the ScrollViewer's viewport. Perhaps all of this is not crucial; I simply want to intercept the ScrollViewer's scroll event, allowing the inner controls to continue responding to the scroll event, while still enabling the ScrollViewer to normally respond to scrolling, with the inner controls not responding to the scroll event.


Can you help me?  In both the on and off states of the switch, ContentScrollViewer can scroll normally, but ZoomHost cannot respond to mouse events.  In another scenario, ContentScrollViewer cannot be scrolled, but ZoomHost receives events normally.
https://github.com/wobushiafa/MangoWidgets.Avalonia/blob/f3996d0e8f405294f7d276f9bedcfc50cd2c8243/MangoWidgets.Avalonia.Sample/Views/TestView.axaml#L16

**Reply by timunie** [reply]
Sorry I cannot develop the control for you. That's up to you. If I needed a ScrwollViewer I'd let it be part of my ControlTemplate.

**Reply by wobushiafa** [reply]
I don't want to define a new control again.  The current software design requires me to implement this unpleasant feature.  Is there really no way to pass the event parameters of one control to another control and let it respond to this event?  Well, maybe this is indeed unreasonable.  Thank you very much for your response.
